packages/doenetml-worker-rust/lib-doenetml-core/src/core/components/component_builder.rs

```diff
@@ -242,6 +229,58 @@ impl ComponentBuilder {
         }
     }
 
+    /// If component `component_idx` extended a prop using the `extend` attribute,
+    /// then add a child to the component corresponding to that prop.
+    ///
+    /// For example, if `<text name="$i"/>`, then we add a child corresponding to `$i.value`
+    /// in the case of `<p extend="$i.value" />` but not to the text component coming from a `$i.value`
+    /// that occurs outside the `extend` attribute.
+    fn add_child_if_prop_from_extend_attribute(&mut self, component_idx: ComponentIdx) {
+        let component = &self.components[component_idx];
+        let new_child = if let Some(Extending::Prop(prop_source)) = component.get_extending() {
+            if !prop_source.from_direct_ref {
+                // the `Extending` was due to specifying a prop inside the `extend` attribute
+
+                let prop_pointer = prop_source.prop_pointer;
+                let referent = &self.components[prop_pointer.component_idx];
+
+                // The creation of the new child mimics `create_component()`
+                // for the case where there is a remaining path corresponding to the prop
+                let new_component_type = referent
+                    .get_prop(prop_pointer.prop_idx)
+                    .unwrap()
+                    .preferred_component_type();
+
+                let mut new_child = ComponentEnum::from_str(new_component_type).unwrap();
+
+                new_child.initialize(
+                    self.components.len(),
+                    Some(component.get_idx()),
+                    Some(Extending::Prop(PropSource {
+                        prop_pointer,
+                        from_direct_ref: true,
+                    })),
+                    HashMap::new(),
+                    None,
+                );
+                Some(new_child)
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+
+        if let Some(new_child) = new_child {
+            // Add the child created to the component as well as the vector of all components
+            let component = &mut self.components[component_idx];
+            let mut children = component.take_children();
+            children.insert(0, UntaggedContent::Ref(new_child.get_idx()));
+            component.set_children(children);
+            self.components.push(new_child);
```

Comment:
This process is complicated. Perhaps a `prepend_child()` utility function could be created?

Reply:
Hah, but that would be a `mut` function! It'll try breaking it into two functions, one `mut` and one not.